regnety/utils/tfrecords_utils.py

```diff
@@ -0,0 +1,288 @@
+
+    image_str = tf.io.encode_jpeg(image_tensor)
+
+    assert len(image_tensor.shape) == 3
+
+    return image_str, height, width
+
+
+def _make_example(
+        image_str: bytes,
+        height: int,
+        width: int, 
+        filepath: str, 
+        label: int, 
+        synset: str) -> tf.train.Example:
+    """
+    Makes a single example from arguments
+
+    Args:
+        image_str: bytes string of image in JPEG RGB format
+        filepath: path to image
+        height: height of image in pixels
+        width: width of image in pixels
+        label: integer denoting label
+        synset: synset string corresponding to image
+
+    Returns:
+        A tf.train.Example having aforementioned attributes
+    """
+
+    try:
+        example = tf.train.Example(
+            features=tf.train.Features(
+                feature={
+                    "image": _bytes_feature(image_str),
+                    "height": _int64_feature(height),
+                    "width": _int64_feature(width),
+                    "filename": _bytes_feature(
+                        bytes(os.path.basename(filepath)).encode("utf8")
+                    ),
+                    "label": _int64_feature(label),
+                    "synset": _bytes_feature(bytes(synset).encode("utf8")),
+                }
+            )
+        )
+    except TypeError:
+        example = tf.train.Example(
+            features=tf.train.Features(
+                feature={
+                    "image": _bytes_feature(image_str),
+                    "height": _int64_feature(height),
+                    "width": _int64_feature(width),
+                    "filename": _bytes_feature(
+                        bytes(os.path.basename(filepath), encoding="utf8")
+                    ),
+                    "label": _int64_feature(label),
+                    "synset": _bytes_feature(bytes(synset, encoding="utf8")),
+                }
+            )
+        )
+    return example
+
+
+def _make_single_tfrecord(
+    chunk_files: List[str],
+    chunk_synsets: List[str],
+    chunk_labels: List[int],
+    output_filepath: str
+):
+
+    """
+    Creates a single TFRecord file having batch_size examples.
+
+    Args:
+        chunk_files: list of filepaths to images
+        chunk_synsets: list of synsets corresponding to images in chunk_files
+        chunk_labels: list of integer labels corresponding to images in
+            chunk_files
+        output_filepath: Output tfrecord file
+
+    Returns None
+    """
+
+    with tf.io.TFRecordWriter(output_filepath) as writer:
+        for i in range(len(chunk_files)):
+            image_str, height, width = _make_image(chunk_files[i])
+            label = chunk_labels[i]
+            synset = chunk_synsets[i]
+
+            example = _make_example(
+                image_str, height, width, chunk_files[i], label, synset
+            )
+
+            writer.write(example.SerializeToString())
+    writer.close()
```

Comment:
Should we print how many TFRecords have been created within some prespecified interval?

Reply:
Actually, I have added the `--log_freq` flag. I don't think it is required for now, maybe I can add it later on if it feels necessary. Is that okay?

Reply:
How are you planning to keep track of the progress then?

Reply:
That didn't cross my mind, sorry about that. I'll add this.

Reply:
> Should we print how many TFRecords have been created within some prespecified interval?

How can we do this elegantly? I thought of two naive solutions, but they both have their issues:

1. Check whether  the predefined interval time has passed sice the last execution of `_make_example` function (it's the function that takes the least time. ) Downside: too many arguments.
2. Same as 1, but use global variables instead of arguments. 

I looked at [sched](https://docs.python.org/3/library/sched.html) and [schedule](https://github.com/dbader/schedule) libraries. None of those provide the ability to pass new arguments on each execution.
Neither of the above are good approaches. Could you please suggest a solution?

Reply:
If the user asks for logging the frequency we log it with respect to an interval we specify. Not letting the user specify this interval is okay IMO. 

@MorganR any particular improvement you'd suggest here?

Reply:
I don't think that the interval being defined by the user or not is the issue. The issue is that when we implement this (using one of the methods mentioned in my earlier comment), the code is a bit less readable. I didn't find any concrete workaround for that. Here's the old [file](https://github.com/AdityaKane2001/regnety/blob/dev/regnety/utils/tfrecords_utils.py) and [here's](https://gist.github.com/AdityaKane2001/ba11ff7f34a8da86e8cb511bc9d03942) the one with the changes. I'll commit these changes if this seems alright.

Reply:
What's stopping us from doing something like so?

```python
if iteration % freq_interval == 0:
    print(f"Serialized {count} TFRecords."
```

`count` could be a global variable here. To my eyes, the code is not anyway unreadable. You can find a complete example here in [this notebook](https://nbviewer.jupyter.org/github/sayakpaul/TF-2.0-Hacks/blob/master/Cats_vs_Dogs_TFRecords.ipynb).

Reply:
I have already implemented this. Code [here](https://github.com/AdityaKane2001/regnety/blob/3d8b1e4636611b205c2b9f09946e18eed311d4f2/regnety/utils/tfrecords_utils.py#L278-L280). I could change this to print after the TFRecord is made.

Reply:
Perfect then.

Reply:
Great, thanks! I'll resolve this.